src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -288,47 +317,88 @@ public void processRecords(List<Record> records, IRecordProcessorCheckpointer ch
-						checkpoint(checkpointer);
-						this.nextCheckpointTimeInMillis = System.currentTimeMillis() + checkpointsInterval;
-					}
 				}
 			}
 
 			// checkpoint if needed
 			if (CheckpointMode.batch.equals(KclMessageDrivenChannelAdapter.this.checkpointMode)) {
 				checkpoint(checkpointer);
 			}
+			else if (CheckpointMode.periodic.equals(KclMessageDrivenChannelAdapter.this.checkpointMode)
+					&& System.currentTimeMillis() > nextCheckpointTimeInMillis) {
+				checkpoint(checkpointer);
+				this.nextCheckpointTimeInMillis = System.currentTimeMillis() + checkpointsInterval;
+			}
 		}
 
-		/**
-		 * Process a single record.
-		 * @param record The record to be processed.
-		 * @param checkpointer the checkpointer to use if the checkpointMode is record
-		 */
 		private void processSingleRecord(Record record, IRecordProcessorCheckpointer checkpointer) {
-			// Convert AWS Record in Spring Message.
-			performSend(prepareMessageForRecord(record, checkpointer), record);
+			performSend(prepareMessageForRecord(record), record, checkpointer);
 
-			// checkpoint if needed
 			if (CheckpointMode.record.equals(KclMessageDrivenChannelAdapter.this.checkpointMode)) {
 				checkpoint(checkpointer);
 			}
+			else if (CheckpointMode.periodic.equals(KclMessageDrivenChannelAdapter.this.checkpointMode)
+					&& System.currentTimeMillis() > nextCheckpointTimeInMillis) {
+				checkpoint(checkpointer);
+				this.nextCheckpointTimeInMillis = System.currentTimeMillis() + checkpointsInterval;
+			}
+		}
+
+		private void processMultipleRecords(List<Record> records, IRecordProcessorCheckpointer checkpointer) {
+			Object payload = records;
+
+			if (KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper != null) {
+				payload = records.stream().map(this::prepareMessageForRecord).collect(Collectors.toList());
+			}
+
+			final List<String> partitionKeys;
+			final List<String> sequenceNumbers;
+			if (KclMessageDrivenChannelAdapter.this.converter != null) {
+				partitionKeys = new ArrayList<>();
+				sequenceNumbers = new ArrayList<>();
+
+				payload = records.stream().map(r -> {
+					partitionKeys.add(r.getPartitionKey());
+					sequenceNumbers.add(r.getSequenceNumber());
+
+					return KclMessageDrivenChannelAdapter.this.converter.convert(r.getData().array());
+				}).collect(Collectors.toList());
+			}
+			else {
+				partitionKeys = null;
+				sequenceNumbers = null;
+			}
+
+			AbstractIntegrationMessageBuilder<?> messageBuilder = getMessageBuilderFactory().withPayload(payload)
+					.setHeader(AwsHeaders.RECEIVED_PARTITION_KEY, partitionKeys)
+					.setHeader(AwsHeaders.RECEIVED_SEQUENCE_NUMBER, sequenceNumbers);
+
+			performSend(messageBuilder, records, checkpointer);
 		}
 
-		private AbstractIntegrationMessageBuilder<Object> prepareMessageForRecord(Record record,
-				IRecordProcessorCheckpointer checkpointer) {
+		private AbstractIntegrationMessageBuilder<Object> prepareMessageForRecord(Record record) {
```

Comment:
I think we still need to perform a conversion for the payload here, similar to what we have in the same method for the
`KinesisMessageDrivenChannelAdapter`: 
```
if (payload instanceof byte[] && KinesisMessageDrivenChannelAdapter.this.converter != null) {
```

The part for the `if (KclMessageDrivenChannelAdapter.this.converter != null) {` in the `processMultipleRecords` needs to be reworked against one or another `payload`. 
We may have an `embeddedHeadersMapper`  and `converter`. Or we may not have an `embeddedHeadersMapper`, but still have a `converter`. I think we need to gather `partitionKeys` and `sequenceNumbers` for both mentioned combinations. When we don't have both options, we just send a `List<Record>`. Let me know if my logic is clear for you!

Reply:
Totally get it, will add the converter tomorrow..

Reply:
Should behave exactly as in `KinesisMessageDrivenChannelAdapter` now. Can you double check?

Some of this code can probably be moved to an abstract parent class in the future.

Reply:
Probably, but for reasons of support we have decided to have separate classes even if they have something common.